text/0436-gamelift-l2.md

```diff
@@ -0,0 +1,983 @@
+
+```ts fix
+import * as gamelift from '@aws-cdk-lib/aws-gamelift';
+
+const fleet = new gamelift.Fleet(this, 'Game server fleet', {
+  build: build
+});
+
+const queue = new gamelift.Queue(this, 'Game Session Queue', {
+  placementTimeout: Duration.seconds(10)
+});
+queue.withDestination(fleet);
+
+const matchmaking = new gamelift.MatchmakingConfigiuration(this, 'Standalone Matchmaking', {
+  requestTimeouts: Duration.seconds(35),
+  ruleSet: MatchmakingRuleSet.fromJsonFile(path.join(__dirname, 'rules.json'))
+});
+matchmaking.withQueue(queue);
+```
+
+See: [FlexMatch integration with GameLift hosting](https://docs.aws.amazon.com/gamelift/latest/flexmatchguide/match-tasks.html)
+in the *Amazon GameLift FlexMatch Developer Guide*.
+
+##### Integrating a queue system
+
+The game session queue is the primary mechanism for processing new game session requests and locating available game servers to host them. Although it is possible to request a new game session be hosted on specific fleet or location.
+
+```ts fixture=with-build
+import * as gamelift from 'aws-cdk-lib/aws-gamelift';
+
+const fleet = new gamelift.Fleet(this, 'Game server fleet', {
+  build: build
+});
+
+const queue = new gamelift.Queue(this, 'Game session queue');
+queue.addDestination(fleet);
+```
+
+or
+
+```ts fixture=with-build
+import * as gamelift from 'aws-cdk-lib/aws-gamelift';
+
+const fleet = new gamelift.Fleet(this, 'Game server fleet', {
+  build: build
+});
+
+const alias = fleet.addAlias('live')
+
+const queue = new gamelift.Queue(this, 'Game session queue');
+queue.addDestination(alias);
+```
+
+See [Setting up GameLift queues for game session placement](https://docs.aws.amazon.com/gamelift/latest/developerguide/realtime-script-uploading.html)
+in the *Amazon GameLift Developer Guide*.
+
+###### Setting notifications
+
+If you're using queues to manage game session placement in your game, you need a way to monitor the status of individual placement requests and take action as appropriate. Implementing event notifications is a fast and efficient method for tracking placement activity. If your game is in production, or in pre-production with high-volume placement activity, you should be using event notifications.
+
+There are two options for setting up event notifications. You can set up an SNS topic and have GameLift publish event notifications on placement activity by referencing the topic ID in a game session queue. Alternatively, you can use Amazon CloudWatch Events, which has a suite of tools available for managing events and taking action on them.
+
+
+```ts fixture=with-build
+import * as gamelift from 'aws-cdk-lib/aws-gamelift';
+
+const fleet = new gamelift.Fleet(this, 'Game server fleet', {
+  build: build
+});
+
+const topic = new sns.Topic(this, 'Topic');
+
+const queue = new gamelift.Queue(this, 'Game session queue', {
+  notification: new SnsDestination(topic)
+});
+queue.addDestination(fleet);
```

Comment:
A question about the GameLift domain: what kinds of messages do these queues store? And what's the difference (if any) between a notification (like sending a message to an SNS topic) and a destination (sending to a `Fleet`)?

Reply:
Here it's not a messaging queue but a player queue so yes on the backend the queue is managing data but it's really a player level Queue used to managed game sessions and placement into different Fleet. The notification system is here in order to received placement update for a player. Imagine a worflow when as a player you are looking for a match, then wait in a Lobby till you match is avalable and then connecting to a gameserver to play.